xdsl/dialects/fsm.py

```diff
@@ -0,0 +1,377 @@
+from __future__ import annotations
+
+from collections.abc import Sequence
+
+from xdsl.dialects.builtin import (
+    ArrayAttr,
+    ContainerOf,
+    DictionaryAttr,
+    FlatSymbolRefAttr,
+    FunctionType,
+    IndexType,
+    IntegerType,
+    StringAttr,
+)
+from xdsl.ir import (
+    Attribute,
+    Block,
+    Operation,
+    Region,
+    SSAValue,
+)
+from xdsl.ir.core import Dialect
+from xdsl.irdl import (
+    AnyAttr,
+    AnyOf,
+    IRDLOperation,
+    attr_def,
+    irdl_op_definition,
+    operand_def,
+    opt_attr_def,
+    opt_operand_def,
+    region_def,
+    var_operand_def,
+    var_result_def,
+)
+from xdsl.traits import (
+    HasParent,
+    IsTerminator,
+    NoTerminator,
+    SymbolOpInterface,
+    SymbolTable,
+)
+from xdsl.utils.exceptions import VerifyException
+
+signlessIntegerLike = ContainerOf(AnyOf([IntegerType, IndexType]))
+
+
+@irdl_op_definition
+class Transition(IRDLOperation):
+    """Represents a transition of a state with a symbol reference
+    of the next state. This op includes an optional `$guard` region with an `fsm.return`
+    as terminator that returns a Boolean value indicating the guard condition of
+    this transition. This op also includes an optional `$action` region that represents
+    the actions to be executed when this transition is taken"""
+
+    name = "fsm.transition"
+
+    guard = region_def()
+
+    action = region_def()
```

Comment:
As far as I understood talking with more experienced people one such things does not really exists! What we do in these cases is basically having a region and keeping it empty, as a "non region"

Reply:
Note that (I think) optional regions were not available in MLIR at some point, so people encoded them as "empty" regions. I think we encountered this when porting the scf dialect for the first time or something. People probably never went back and actually made the regions optional, since that influences quite a lot of the repo, so there are these weird left overs at times.